modules/dnn/src/model.cpp

```diff
@@ -197,27 +197,78 @@ void Model::predict(InputArray frame, OutputArrayOfArrays outs) const
 }
 
 
+class ClassificationModel_Impl : public Model::Impl
+{
+public:
+    virtual ~ClassificationModel_Impl() {}
+    ClassificationModel_Impl() : Impl() {}
+    ClassificationModel_Impl(const ClassificationModel_Impl&) = delete;
+    ClassificationModel_Impl(ClassificationModel_Impl&&) = delete;
+
+    void setEnableSoftmaxPostProcessing(bool enable)
+    {
+        applySoftmax = enable;
+    }
+
+    bool getEnableSoftmaxPostProcessing() const
+    {
+        return applySoftmax;
+    }
+
+private:
+    bool applySoftmax = false;
+};
+
+ClassificationModel::ClassificationModel()
+    : Model()
+{
+    // nothing
+}
+
 ClassificationModel::ClassificationModel(const String& model, const String& config)
-    : Model(model, config)
+    : ClassificationModel(readNet(model, config))
 {
     // nothing
 }
 
 ClassificationModel::ClassificationModel(const Net& network)
     : Model(network)
 {
-    // nothing
+    impl = makePtr<ClassificationModel_Impl>();
+    impl->initNet(network);
+}
+
+ClassificationModel& ClassificationModel::setEnableSoftmaxPostProcessing(bool enable)
+{
+    CV_Assert(impl != nullptr && impl.dynamicCast<ClassificationModel_Impl>() != nullptr);
+    impl.dynamicCast<ClassificationModel_Impl>()->setEnableSoftmaxPostProcessing(enable);
+    return *this;
+}
+
+bool ClassificationModel::getEnableSoftmaxPostProcessing() const
+{
+    CV_Assert(impl != nullptr && impl.dynamicCast<ClassificationModel_Impl>() != nullptr);
+    return impl.dynamicCast<ClassificationModel_Impl>()->getEnableSoftmaxPostProcessing();
 }
 
 std::pair<int, float> ClassificationModel::classify(InputArray frame)
 {
+    CV_Assert(impl != nullptr && impl.dynamicCast<ClassificationModel_Impl>() != nullptr);
+
     std::vector<Mat> outs;
     impl->processFrame(frame, outs);
     CV_Assert(outs.size() == 1);
 
+    Mat out = outs[0].reshape(1, 1);
+
+    if (getEnableSoftmaxPostProcessing())
+    {
+        cv::dnn::softmax(out, out);
+    }
+
     double conf;
-    cv::Point maxLoc;
-    minMaxLoc(outs[0].reshape(1, 1), nullptr, &conf, nullptr, &maxLoc);
+    Point maxLoc;
+    cv::minMaxLoc(out, nullptr, &conf, nullptr, &maxLoc);
     return {maxLoc.x, static_cast<float>(conf)};
 }
```

Comment:
Implementation details should be moved into `ClassificationModel_Impl::classify()`.
`ClassificationModel::classify()` should perform as a thin wrapper.

Reply:
@alalek Is the same true for ClassificationModel::softmax()?

```
class ClassificationModel_Impl : public Model::Impl
{
public:
    void softmax(InputArray inblob, OutputArray outblob)
    {
    }
}

void ClassificationModel::softmax(InputArray inblob, OutputArray outblob)
{
    CV_Assert(impl != nullptr && impl.dynamicCast<ClassificationModel_Impl>() != nullptr);
    impl.dynamicCast<ClassificationModel_Impl>()->softmax(inblob, outblob);
}
```

Reply:
We don't need `softmax()` method in public `ClassificationModel`. Lets keep public APIs clear.
We will discuss how to provide dedicated `softmax()` processing a bit later.

There is no such issue with internal `ClassificationModel_Impl` as we don't have compatibility obligations there.

Reply:
I understand. Remove softmax() function from Classification class.